File: src/arrays/encodeanddecodestrings/string_encode_and_decode.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func encode(strs []string) string {

	// For any string operation always check empty or null
	if len(strs) == 0 {
		return ""
	}

	var encodedString = ""

	// use ::  as delimiter and ,, as collector
	for _, str := range strs {
		if encodedString != "" {
			encodedString += ",,"
		}
		length := len(str)
		encodedString += strconv.Itoa(length) + "::" + str
	}

	return encodedString
}

func decode(str string) []string {

	// Always check empty
	if str == "" {
		return []string{}
	}

	// Separate the collectors
	encodedStrings := strings.Split(str, ",,")

	// Dynamic array with the same size
	decodedStrings := make([]string, len(encodedStrings))

	// Rebuild the previous array
	for index, encodedString := range encodedStrings {
		parts := strings.Split(encodedString, "::")
		decodedString := ""
		if len(parts) > 0 {
			decodedString = parts[1]
		}
		decodedStrings[index] = decodedString
	}

	return decodedStrings
}

func main() {

	//Expect the same string
	fmt.Println(decode(encode([]string{"neet", "code", "love", "you"})))

	//Expect the same string
	fmt.Println(decode(encode([]string{"we", "say", ":", "yes"})))

	//Expect the same string
	fmt.Println(decode(encode([]string{""})))

	//Expect the same string
	fmt.Println(decode(encode([]string{})))
}

File: src/arrays/encodeanddecodestrings/string_encode_and_decode.rs

fn encode(strs : &[&str]) -> String {

    // Always check for empty in arrays
    if  strs.len() == 0 {
        return String::from("");
    }

    let mut encoded_string : String = String::from("");

    for str in strs {
        // Set a collector
        if encoded_string != "" {
            encoded_string = encoded_string + ",,";
        }

        // Create the delimiter and join
        let str_len = str.chars().count();
        encoded_string = encoded_string + &str_len.to_string() + "::" + str;
    }

    return encoded_string;
}

fn decode(str: String) -> Vec<String> {
    // Always check for empty
    if str .is_empty() {
        return Vec::new();
    }

    // Get an iterator of the encoded strings after being split
    let encoded_strings =  str.split(",,");

    // Stream through the iter and consume the values
    encoded_strings.into_iter().map(|encoded_string| {
        let mut value = String::from("");

        // Split the delimiter and set it
        let parts: Vec<_> = encoded_string.split("::").collect();

        if  parts.len() > 0{
            value = String::from(parts[1]);
        }
        return  value;
    }).collect()
}

fn main(){
    //Expect the same string
    println!("{:?}", decode(encode(&["neet", "code", "love", "you"])));

    //Expect the same string
    println!("{:?}", decode(encode(&["we", "say", ":", "yes"])));

    //Expect the same string
    println!("{:?}", decode(encode(&[""])));

    //Expect the same string
    println!("{:?}", decode(encode(&[])));
}
